include/cuco/detail/bloom_filter/bloom_filter_impl.cuh

```diff
@@ -0,0 +1,359 @@
+
+  template <class InputIt, class StencilIt, class Predicate, class OutputIt>
+  __host__ void test_if(InputIt first,
+                        InputIt last,
+                        StencilIt stencil,
+                        Predicate pred,
+                        OutputIt output_begin,
+                        cuda::stream_ref stream) const
+  {
+    this->test_if_async(first, last, stencil, pred, output_begin, stream);
+    stream.wait();
+  }
+
+  template <class InputIt, class StencilIt, class Predicate, class OutputIt>
+  __host__ void test_if_async(InputIt first,
+                              InputIt last,
+                              StencilIt stencil,
+                              Predicate pred,
+                              OutputIt output_begin,
+                              cuda::stream_ref stream) const noexcept
+  {
+    auto const num_keys = cuco::detail::distance(first, last);
+    if (num_keys == 0) { return; }
+
+    auto constexpr block_size = cuco::detail::default_block_size();
+    auto const grid_size =
+      cuco::detail::grid_size(num_keys, 1, cuco::detail::default_stride(), block_size);
+
+    detail::test_if_n<block_size><<<grid_size, block_size, 0, stream.get()>>>(
+      first, num_keys, stencil, pred, output_begin, *this);
+  }
+
+  [[nodiscard]] __host__ __device__ word_type* data() noexcept { return words_; }
+
+  [[nodiscard]] __host__ __device__ word_type const* data() const noexcept { return words_; }
+
+  [[nodiscard]] __host__ __device__ extent_type block_extent() const noexcept
+  {
+    return num_blocks_;
+  }
+
+  [[nodiscard]] __host__ __device__ uint32_t pattern_bits() const noexcept { return pattern_bits_; }
+
+  [[nodiscard]] __host__ __device__ hasher hash_function() const noexcept { return hash_; }
+
+  // TODO
+  // [[nodiscard]] __host__ double occupancy() const;
+  // [[nodiscard]] __host__ double expected_false_positive_rate(size_t unique_keys) const
+  // [[nodiscard]] __host__ __device__ static uint32_t optimal_pattern_bits(size_t num_blocks)
+  // template <typename CG, cuda::thread_scope NewScope = thread_scope>
+  // [[nodiscard]] __device__ constexpr auto make_copy(CG const& group, word_type* const
+  // memory_to_use, cuda_thread_scope<NewScope> scope = {}) const noexcept;
+
+ private:
+  template <class HashValue>
+  __device__ size_type block_idx(HashValue hash_value) const
+  {
+    // TODO use fast_int modulo
+    return hash_value % num_blocks_;
+  }
+
+  // we use the LSB bits of the hash value to determine the pattern bits for each word
+  template <class HashValue>
+  __device__ auto pattern(HashValue hash_value) const
+  {
+    cuda::std::array<word_type, block_words> pattern{};
+    auto constexpr word_bits           = sizeof(word_type) * CHAR_BIT;
+    auto constexpr bit_index_width     = cuda::std::bit_width(word_bits - 1);
+    word_type constexpr bit_index_mask = (word_type{1} << bit_index_width) - 1;
+
+    auto const bits_per_word = pattern_bits_ / block_words;
```

Comment:
```suggestion
    auto const pattern_bits_per_word = pattern_bits_ / words_per_block;
```

even better to have two member functions: `pattern_bits_per_word()` and `pattern_bits_remainder()`